### PR TITLE
XLOOKUP function

### LINKED ISSUE #1458: Add XLOOKUP function
### Context

XLOOKUP is a popular and useful function.

Specification: https://support.microsoft.com/en-us/office/xlookup-function-b7fd680e-6d10-43e6-84f9-88eae8bf5929

Originally discussed in https://github.com/handsontable/hyperformula/discussions/1064 and the starting point for the implementation done in https://github.com/handsontable/hyperformula/pull/1414 by @selimyoussry. Let's accept the above PR and finish it to make XLOOKUP available in HyperFormula as a built-in function.

### PR DESCRIPTION
**Pull request by sequba**
### Context

- [x] merge PR by @selimyoussry 
- [x] add more unit tests for XLOOKUP including several edge cases (total: 109 unit tests for XLOOKUP)
- [x] adjust the implementation to meet the criteria specified by the tests
- [x] add changelog entry
- [x] add entry to the built-in functions page

### How did you test your changes?

- linter passed
- 109 new unit tests for different scenarios of XLOOKUP
- all unit tests passed

### Types of changes

- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [x] New feature or improvement (a non-breaking change that adds functionality)
- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
Fixes #1458 

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to HyperFormula](https://hyperformula.handsontable.com/guide/contributing.html) and I confirm that my code follows the code style of this project.
- [ ] I have signed the [Contributor License Agreement](https://goo.gl/forms/yuutGuN0RjsikVpM2).
- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [ ] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [x] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

File: src/Lookup/AdvancedFind.ts [hunk 1]
```diff
@@ -11,59 +11,93 @@ import {
...
 
 export abstract class AdvancedFind {
   protected constructor(
     protected dependencyGraph: DependencyGraph
   ) {
   }
 
-  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, rangeValue: SimpleRangeValue): number {
-    let values: InternalScalarValue[]
+  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, rangeValue: SimpleRangeValue, { returnOccurence }: AdvancedFindOptions = { returnOccurence: 'first' }): number {
```
Comment (budnix) [inline src/Lookup/AdvancedFind.ts:26]: a typo :) 
```suggestion
  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, rangeValue: SimpleRangeValue, { returnOccurence }: AdvancedFindOptions = { returnOccurrence: 'first' }): number {
```

File: src/Lookup/ColumnIndex.ts [hunk 2]
```diff
@@ -107,16 +107,20 @@ export class ColumnIndex implements ColumnSearchStrategy {
...
   }
 
-  /*
-   * WARNING: Finding lower/upper bounds in unordered ranges is not supported. When ordering === 'none', assumes matchExactly === true
-   */
-  public find(searchKey: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, { ordering, matchExactly }: SearchOptions): number {
-    const handlingDuplicates = matchExactly === true ? 'findFirst' : 'findLast'
-    const resultUsingColumnIndex = this.findUsingColumnIndex(searchKey, rangeValue, handlingDuplicates)
-    return resultUsingColumnIndex !== undefined ? resultUsingColumnIndex : this.binarySearchStrategy.find(searchKey, rangeValue, { ordering, matchExactly })
+  public find(searchKey: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, { ordering, ifNoMatch, returnOccurence }: SearchOptions): number {
```
Comment (budnix) [inline src/Lookup/ColumnIndex.ts:110]:
```suggestion
  public find(searchKey: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, { ordering, ifNoMatch, returnOccurrence }: SearchOptions): number {
```

File: src/Lookup/ColumnIndex.ts [hunk 3]
```diff
@@ -107,16 +107,20 @@ export class ColumnIndex implements ColumnSearchStrategy {
...
+    if (returnOccurence == null) {
+      returnOccurence = ordering === 'none' ? 'first' : 'last'
+    }
+
+    const resultUsingColumnIndex = this.findUsingColumnIndex(searchKey, rangeValue, returnOccurence)
+    return resultUsingColumnIndex !== undefined ? resultUsingColumnIndex : this.binarySearchStrategy.find(searchKey, rangeValue, { ordering, ifNoMatch, returnOccurence })
   }
 
-  private findUsingColumnIndex(key: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, handlingDuplicates: 'findFirst' | 'findLast'): Maybe<number> {
+  private findUsingColumnIndex(key: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, returnOccurence: 'first' | 'last'): Maybe<number> {
```
Comment (budnix) [inline src/Lookup/ColumnIndex.ts:123]:
```suggestion
  private findUsingColumnIndex(key: RawNoErrorScalarValue, rangeValue: SimpleRangeValue, returnOccurrence: 'first' | 'last'): Maybe<number> {
```

File: src/Lookup/ColumnIndex.ts [hunk 4]
```diff
@@ -135,15 +139,15 @@ export class ColumnIndex implements ColumnSearchStrategy {
       return undefined
     }
 
-    const rowNumber = ColumnIndex.findRowBelongingToRange(valueIndexForTheKey, range, handlingDuplicates)
+    const rowNumber = ColumnIndex.findRowBelongingToRange(valueIndexForTheKey, range, returnOccurence)
     return rowNumber !== undefined ? rowNumber - range.start.row : undefined
   }
 
-  private static findRowBelongingToRange(valueIndex: ValueIndex, range: AbsoluteCellRange, handlingDuplicates: 'findFirst' | 'findLast'): Maybe<number> {
+  private static findRowBelongingToRange(valueIndex: ValueIndex, range: AbsoluteCellRange, returnOccurence: 'first' | 'last'): Maybe<number> {
```
Comment (budnix) [inline src/Lookup/ColumnIndex.ts:146]:
```suggestion
  private static findRowBelongingToRange(valueIndex: ValueIndex, range: AbsoluteCellRange, returnOccurrence: 'first' | 'last'): Maybe<number> {
```

File: src/Lookup/ColumnIndex.ts [hunk 5]
```diff
@@ -158,8 +162,8 @@ export class ColumnIndex implements ColumnSearchStrategy {
   }
 
 
-  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, range: SimpleRangeValue): number {
-    return this.binarySearchStrategy.advancedFind(keyMatcher, range)
+  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, range: SimpleRangeValue, options: AdvancedFindOptions = { returnOccurence: 'first' }): number {
```
Comment (budnix) [inline src/Lookup/ColumnIndex.ts:165]:
```suggestion
  public advancedFind(keyMatcher: (arg: RawInterpreterValue) => boolean, range: SimpleRangeValue, options: AdvancedFindOptions = { returnOccurrence: 'first' }): number {
```

File: src/Lookup/SearchStrategy.ts [hunk 6]
```diff
@@ -16,7 +16,12 @@ import {ColumnIndex} from './ColumnIndex'
 
 export interface SearchOptions {
   ordering: 'asc' | 'desc' | 'none',
-  matchExactly?: boolean,
+  ifNoMatch: 'returnLowerBound' | 'returnUpperBound' | 'returnNotFound',
+  returnOccurence?: 'first' | 'last',
+}
+
+export interface AdvancedFindOptions {
+  returnOccurence?: 'first' | 'last',
```
Comment (budnix) [inline src/Lookup/SearchStrategy.ts:24]:
```suggestion
  returnOccurrence?: 'first' | 'last',
}

export interface AdvancedFindOptions {
  returnOccurrence?: 'first' | 'last',
```

**Review by budnix — CHANGES_REQUESTED** [review]
I'm not an expert in HF but it looks good to me. I only found a typo that worth to fix.

File: src/Lookup/AdvancedFind.ts (hunk 1 above)
Comment (sequba) [reply src/Lookup/AdvancedFind.ts]: Thanks @budnix, great catch. I changed every occurrence of "occurence" (single r).

**Review by budnix — APPROVED** [review]
